module/mfrc522_parser.c

```diff
@@ -0,0 +1,177 @@
+#include "mfrc522_command.h"
+
+#include <linux/string.h>
+#include <linux/slab.h>
+#include <linux/types.h>
+
+#include "mfrc522_parser.h"
+
+#define MFRC522_SEPARATOR ":"
+#define MFRC522_CMD_AMOUNT 4
+#define MFRC522_MAX_INPUT_LEN 255
+#define MFRC522_MAX_PARAMETER_AMOUNT 2
+
+struct driver_command {
+	const char *input;
+	u8 parameter_amount;
+	u8 cmd;
+};
+
+static const struct driver_command commands[MFRC522_CMD_AMOUNT] = {
+	{ .input = "mem_write",
+	  .parameter_amount = 2,
+	  .cmd = MFRC522_CMD_MEM_WRITE },
+	{ .input = "mem_read",
+	  .parameter_amount = 0,
+	  .cmd = MFRC522_CMD_MEM_READ },
+	{ .input = "gen_rand_id",
+	  .parameter_amount = 0,
+	  .cmd = MFRC522_CMD_GEN_RANDOM },
+	{ .input = "version",
+	  .parameter_amount = 0,
+	  .cmd = MFRC522_CMD_GET_VERSION },
+};
+
+/**
+ * Get the command associated with a command name
+ *
+ * @param token Command name found during parsing
+ *
+ * @return Return a pointer to a command declared in the commands array, or NULL if no
+ *         command matched the input
+ */
+static const struct driver_command *find_cmd_from_token(const char *token)
+{
+	size_t i;
+
+	for (i = 0; i < MFRC522_CMD_AMOUNT; i++)
+		if (!strcmp(token, commands[i].input))
+			return &commands[i];
+
+	return NULL;
+}
+
+/**
+ * Parse a command with multiple arguments
+ *
+ * @param cmd Command struct to fill up
+ * @param input Remaining user input, mutable and allocated via kmalloc() in mfrc522_parse()
+ * @param ref_cmd Reference command returned by find_cmd_from_token()
+ *
+ * @return 0 on success, a negative number otherwise
+ */
+static int parse_multi_arg(struct mfrc522_command *cmd, char *input,
+			   const struct driver_command *ref_cmd)
+{
+	// We only enter this function if arguments have been given to the input
+	u8 parameter_amount = 0;
+	char *token;
+	char *extra_data;
+	u8 extra_data_len;
+	int ret;
+
+	while (input) {
+		token = strsep(&input, MFRC522_SEPARATOR);
+		parameter_amount++;
+
+		// The first parameter is the length of the extra data
+		if (parameter_amount == 1) {
+			ret = kstrtou8(token, 10, &extra_data_len);
+			if (ret == -EINVAL) {
+				pr_err("[MFRC522] Invalid parameter for Data length: Expected number but got %s\n",
+				       token);
+				return ret;
+			}
+
+			if (ret == -ERANGE) {
+				pr_err("[MFRC522] Invalid parameter for Data length: Expected Unsigned Byte (0 - 255) but got %s\n",
+				       token);
+				return ret;
+			}
+
+			if (extra_data_len > MFRC522_MAX_DATA_LEN) {
+				pr_err("[MFRC522] Invalid parameter for Data length: Length %d is too important (max length: 25)\n",
+				       extra_data_len);
+				return -1;
+			}
+		} else if (parameter_amount == 2) {
+			// The second parameter is the extra data
+			extra_data = token;
+		} else {
+			// Any parameters after the second one is an error
+			pr_err("[MFRC522] Invalid number of parameters: Got at least 3, which is more than the maximum of %d",
+			       MFRC522_MAX_PARAMETER_AMOUNT);
+			return -1;
+		}
```

Comment:
Because if there is a byte whose value is 58 (':' in ascii), strsep will wrongly cut the ``extra_data``. A solution would be to assign extra_data at line 99: ``extra_data=input`` and then do something like ``input+=ret`` (with some out of bound check)

Reply:
@n1tram1 An idea ?